tools/pyvmi/libvmi.py

```diff
@@ -0,0 +1,665 @@
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.destroy()
+
+    def destroy(self):
+        if self.vmi:
+            status = lib.vmi_destroy(self.vmi)
+            check(status)
+        self.opaque_vmi = None
+        self.vmi = None
+
+    def get_library_arch(self):
+        arch = lib.vmi_get_library_arch()
+        return VMIArch(arch)
+
+    def get_rekall_path(self):
+        value = lib.vmi_get_rekall_path(self.vmi)
+        if value == ffi.NULL:
+            return None
+        return ffi.string(value).decode()
+
+    # memory translations
+    def translate_kv2p(self, vaddr):
+        paddr = ffi.new("addr_t *")
+        status = lib.vmi_translate_kv2p(self.vmi, vaddr, paddr)
+        check(status)
+        return paddr[0]
+
+    def translate_uv2p(self, vaddr, pid):
+        paddr = ffi.new("addr_t *")
+        status = lib.vmi_translate_uv2p(self.vmi, vaddr, pid, paddr)
+        check(status)
+        return paddr[0]
+
+    def translate_ksym2v(self, symbol):
+        vaddr = ffi.new("addr_t *")
+        status = lib.vmi_translate_ksym2v(self.vmi, symbol.encode(), vaddr)
+        check(status)
+        return vaddr[0]
+
+    def translate_sym2v(self, ctx, symbol):
+        vaddr = ffi.new("addr_t *")
+        status = lib.vmi_translate_sym2v(self.vmi, ctx, symbol.encode(), vaddr)
+        check(status)
+        return vaddr[0]
+
+    def translate_v2sym(self, ctx, addr):
+        symbol = lib.vmi_translate_v2sym(self.vmi, ctx, addr)
+        if symbol == ffi.NULL:
+            raise LibvmiError('VMI_FAILURE')
+        return ffi.string(symbol).decode()
+
+    def translate_v2ksym(self, ctx, addr):
+        symbol = lib.vmi_translate_v2ksym(self.vmi, ctx, addr)
+        if symbol == ffi.NULL:
+            raise LibvmiError('VMI_FAILURE')
+        return ffi.string(symbol).decode()
+
+    def pid_to_dtb(self, pid):
+        dtb = ffi.new('addr_t *')
+        status = lib.vmi_pid_to_dtb(self.vmi, pid, dtb)
+        check(status)
+        return dtb[0]
+
+    def dtb_to_pid(self, dtb):
+        pid = ffi.new("vmi_pid_t *")
+        status = lib.vmi_dtb_to_pid(self.vmi, dtb, pid)
+        check(status)
+        return pid[0]
+
+    def pagetable_lookup(self, dtb, vaddr):
+        paddr = ffi.new("addr_t *")
+        status = lib.vmi_pagetable_lookup(self.vmi, dtb, vaddr, paddr)
+        check(status)
+        return paddr[0]
+
+    def pagetable_lookup_extended(self, dtb, vaddr):
+        page_info = ffi.new("page_info_t *")
+        status = lib.vmi_pagetable_lookup_extended(self.vmi, dtb, vaddr, page_info)
+        check(status)
+        return page_info
+
+    # read
+    def read(self, ctx, count):
+        buffer = ffi.new("char[]", count)
+        bytes_read = ffi.new("size_t *")
+        status = lib.vmi_read(self.vmi, ctx, count, buffer, bytes_read)
+        check(status)
+        # transform into Python bytes
+        buffer = ffi.buffer(buffer, bytes_read[0])[:]
```

Comment:
You can use the [`ffi.string`](http://cffi.readthedocs.io/en/latest/ref.html#ffi-string-ffi-unpack) interface here which is more neat. In this way you don't have to slice-copy the buffer.

Reply:
replaced, thanks.